pkg/api/controllers/catalog/catalog.go

```diff
@@ -142,3 +224,47 @@ func cleanupPath(dir string, files []string, valid map[string]bool) int {
 	}
 	return count
 }
+
+func checkCatalogUpdate(m *manager.Manager, catalog *v3.Catalog, cmt *manager.CatalogInfo) (runtime.Object, error) {
+	hashPath := helmlib.CatalogSHA256Hash(catalog)
+	path := filepath.Join(helmlib.CatalogCache, hashPath)
+	if _, err := os.Stat(path); err == nil {
+		if git.IsValid(catalog.Spec.URL) {
+			// if path exist, check if the updated commit is different from the local cache commit
+			headCommit, err := git.HeadCommit(path)
+			if err != nil {
+				return nil, err
+			}
+			if !manager.IsUpToDate(headCommit, catalog) {
+				return newForceUpdateCatalog(m, catalog, cmt)
+			}
+		} else {
+			// if it's not a git url, update as helm index catalog
+			return newForceUpdateCatalog(m, catalog, cmt)
+		}
+	} else if os.IsNotExist(err) {
+		return newForceUpdateCatalog(m, catalog, cmt)
+	} else {
+		return nil, err
+	}
+	return nil, nil
+}
+
+func newForceUpdateCatalog(m *manager.Manager, obj *v3.Catalog, cmt *manager.CatalogInfo) (runtime.Object, error) {
+	catalogType := manager.GetCatalogType(cmt)
+	commit, helm, err := helmlib.NewForceUpdate(obj)
```

Comment:
Instead of calling helmlib NewForceUpdate it is probably better to call NewNoUpdate and then UpdateGit with a specific commit. There are likely some updates needed to helmlib & git packages to make this work well.

Reply:
I think the `helmlib.NewForceUpdate` is literally doing the same work as we want,  call NewNoUpdate and then make `updateGit` or `updateIndex` by validating the URL.

Reply:
The difference is that we are checking to see if it is a valid git url when we have previously already downloaded and verified it is a git ur. This is a problem for the current helmlib code. If the repo had previously been checked out as git but there is some temporary network issue we would start reporting errors that index.yaml can not be downloaded, which we should not be trying to do.

Reply:
I think this is a known issue of the `helmlib` and we should fix it, but this doesn't apply to HA mode only, both current and single Rancher mode will have this issue, maybe we should put a separate fixing PR for this.

Reply:
Sure, it is maybe a one line change.

Reply:
Here is a PR for helm which fixes HA mode: rancher/rancher/pull/20198

Reply:
updated in [here](https://github.com/rancher/rancher/pull/19856/files#diff-5254979f18a6777e06795019dbab72eaR109) will add local git cache check. And we can add https://github.com/rancher/rancher/pull/20198 enhancement in separately as they are not necessary to be linked.

Reply:
They are linked, as I suggested earlier there is more than one way to approach this. In #20198 the behavior of helm.New is updated to use the exact commit that is provided in the catalog.